### PR TITLE
batches: support distinct GH App installs for commit signing across namespaces 

### PR DESCRIPTION
**Pull request by courier-new**
Closes https://github.com/sourcegraph/sourcegraph/issues/53280.

Adds support for installing a Batches domain GitHub App on multiple accounts and then locating the appropriate installation by `repo namespace == account name` in order to authenticate the `ChangesetSource` for duplicating and signing commits.

### Walkthrough

I'm the master of producing chunky PRs lately so here's another code walkthrough to ease the burden of reviewing this. Remember that more than half the files changed are just for stupid `logger` plumbing and nothing else. 🙈 We'll go back to front again.

**new migrations** - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-3e570c17b9e46905c65e71bf5c71a19af435640247ed9d15ca296e1ed1f1d203)]

- The first migration adds additional metadata about a GitHub App Installation to the `github_app_installs` table. It also creates an index based on the `account_login`, since that's how we'll look up the right install for a given commit signing interaction.
- The second migration adds a unique constraint on for `app_id` and `installation_id`. This enables us to easily update records with the same SQL `INSERT` statement we use to add a new one.

`github_apps/store` - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-e8fe19664bf9f7c5e69983ecad802bdf6610083828a4d3faee9eddf430ab068f)]

- `Install` now takes a fully formed `ghtypes.GitHubAppInstallation` as an input and performs an upsert (insert or update) of the record.
- `GetLatestInstallID` has been replaced with the more relevant `GetInstallID`, which looks up installations by app ID and account/namespace name and returns the installation ID of the latest record that matches.
- `BulkInstall` was removed because it was easier to just loop and upsert individual records.
- `SyncInstallations` is a new method that performs the logic that was originally in the installation backfill worker. This allows us to tap into it in another place. The logic has been enhanced slight as well:
  - We now delete all Installations for an App if that App was not reachable on GitHub.
  - We now add (upsert) records for all Installations returned from the remote request, including existing ones. It upserts them with all the additional metadata.

`worker/installation_backfill` - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-e96fa3fa28e1267bca5210ed98f7d70e226ea2d67c77d03b6406546e7209fbec)]

- The worker routine has been simplified to mostly just preparing a GitHub client and then calling `store.SyncInstallations`. The test has largely been migrated to the `store` package tests instead.

`batches/sources` - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-bb8e7698f8e9e949afb9bb20f0015f11bb1f570ff02082e9ac39e539d828dfe4)]

- `withGitHubAppAuthenticator` now takes `account` as an additional argument and uses it to look up the installation for both the right app ID and account name when creating the installation authenticator.

`batches.graphql` - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-e81543ff15f2f5ee599339869bf4adacf439d298eb13eb80909c1aba6404030f)]

- I replaced our `GitHubAppConfiguration` type with the same `GitHubApp` type from the GH Apps GraphQL schema, since we're increasingly needing more and more properties of the App in the Batches space and it was beginning to feel redundant. Now the `commitSigningConfiguration` field will resolve to a full `GitHubApp` rather than our shallow copy of one.

**anything in** `frontend/internal/batches`

- Almost everything here is literally me just passing a `logger` or maybe an `EnterpriseDB` around. It was extremely painful to write, but hopefully less painful to skim. 🙂
- The only exception is [`code_host.go`](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-3001ae302baaee8046a67c6768baebfe06386a819aad4f52ea98df7a4af5df39), where we now nest a `githubapp.NewGitHubAppResolver` for the commit signing configuration.

`graphqlbackend.graphqlbackend.go` - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-f7ca4628497132bb9d98b2fda8508ac30b2c64ac571c67547a6dd82ee3c94c51)]

- Because the Batches GraphQL Schema now depends on the GitHub Apps Schema also being available, I updated `NewSchemaWithBatchChangesResolver` to provide both as resolvers, and updated `NewSchema` to accept a slice of resolvers instead of just one. This had a bit of a ripple effect in that I had to update everyone who calls this method to call it with a slice now. 😅

`githubappauth/resolver.go` - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-7fdd07cc0cf6d14dab030dc49b6fe2e1ac200c9e0abb0439179f5491d5cd705d)]

- Originally, to resolve the `Installations` field on a `GitHubApp`, this resolver would requery GitHub itself for the Installations every time. Now we take the Installations we have stored in the DB to resolve this field faster.
- We also opportunistically kick of `store.SyncInstallations` whenever `Installations` is called, so the set of Installations for the `GitHubApp` will be updated immediately after if they are out-of-date. This has the added benefit of providing us a way to trigger a "manual sync" of GitHub App Installations if needed.
- This sync process is offloaded to a goroutine so as not to block the GraphQL request from completing.
- Resolving `Installations` is also now guarded behind a `sync.Once` mechanism to ensure the function is only executed once for a given GraphQL request.

`githubappauth/middleware.go` [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-89757e87b4fa8f63065ad8da9d5ebd8e496fb6ee9e6b6b0afcc12ab36586fc58)]

- The `/setup` handler will now fetch the whole GitHub App Installation before saving it in the DB, to ensure we can populate all the necessary metadata.

`GitHubAppPage` - [[code](https://github.com/sourcegraph/sourcegraph/pull/53212/files?w=1#diff-146fef419c6a36935a86088ecb42710fd57f6a2615fc96c77bc039cda07f5e8c)]

- This page now has a configurable header breadcrumb.
- Rather than showing the warning banner about being un-editable if it's rendered for a Batches App, now it shows just the parts of the page that are relevant to Batch Changes, namely, the basic App details and the list of Installations registered on Sourcegraph.
- This page is now accessible for Batches domain Apps at `/site-admin/batch-changes/github-apps/:id`.

### Demo

Prior to this PR, we would just pretend that whatever GitHub App installation we had would work for any repositories on that code host. That would mean that publishing changesets on repositories from separate orgs wouldn't just not produce signed commits, it would actually just _outright fail to publish_.

In the demo, I start out with a new GitHub App, which I only install to my personal user account namespace. I test that when I publish two changesets, one to a repo in my user namespace and one to a repo in the `sourcegraph-testing` namespace, they both succeed to publish. In this case, the `sourcegraph-testing` one just falls back on normal batch change user credentials, yielding an unsigned commit but one which is authored by me. However, the `courier-new` one is able to authenticate with as the GH App installation for that account and duplicate the commit, yielding the signed one!

Then, I make the GH App publish and install it for `sourcegraph-testing`. I modify the branch name for my batch spec to trigger new changesets to be published and publish the two again. This time, both succeed *and* have signed commits.

https://github.com/sourcegraph/sourcegraph/assets/8942601/965d02dd-5e2d-4d5c-87e5-b0ef39bad14a

## Test plan

Updated unit tests where applicable. Manual testing of the whole flow.

### CODE REVIEW

**Review by st0nebraker — APPROVED** [review]
This is great! Thank you Kelli! I manually tested and see it all working 😄 

I am noticing with multiple installs, the 2nd nodes "View in GitHub" link is broken. Do you see this on your end?

https://github.com/sourcegraph/sourcegraph/assets/59381432/740bd8de-6c42-40d9-9983-0486d9c1198a

File: enterprise/internal/github_apps/store/store_test.go [hunk 1]
```diff
@@ -515,12 +517,28 @@ func TestInstallGitHubApp(t *testing.T) {
 	}
 
 	id, err := store.Create(ctx, app)
+	fmt.Printf("github app id: %v\n", id)
```
Comment (pjlast) [inline enterprise/internal/github_apps/store/store_test.go:520]: A stray printf :D

**Review by pjlast — APPROVED** [review]
Not much to add, just see comment on table constraint 🙏

File: enterprise/internal/github_apps/store/store.go [hunk 2]
```diff
@@ -424,3 +446,105 @@ func (s *gitHubAppsStore) BulkRemoveInstallations(ctx context.Context, id int, i
...
+			}
+		}
+
+		return errs
+	}
+
+	var remoteInstallsMap = make(map[int]struct{}, len(remoteInstallations))
+	for _, in := range remoteInstallations {
+		remoteInstallsMap[int(*in.ID)] = struct{}{}
+	}
```
[the remark this answers is not in the document]
Comment (courier-new) [reply enterprise/internal/github_apps/store/store.go:487]: Ah you're right, no I forgot to take it out. 🙂 It was used in Bolaji's original worker, but not after my slight refactor. Thanks for spotting!

File: migrations/frontend/squashed.sql [hunk 3]
```diff
@@ -5563,6 +5569,9 @@ ALTER TABLE ONLY temporary_settings
 ALTER TABLE ONLY temporary_settings
     ADD CONSTRAINT temporary_settings_user_id_key UNIQUE (user_id);
 
+ALTER TABLE ONLY github_app_installs
+    ADD CONSTRAINT unique_app_install UNIQUE (app_id, installation_id);
+
```
[the remark this answers is not in the document]
Comment (courier-new) [reply migrations/frontend/squashed.sql:5574]: Oh right, good call!

[the remark this answers is not in the document]
Comment (courier-new) [reply migrations/frontend/squashed.sql:5574]: Oh actually, on second thought, I think it's actually okay with just `app_id`, `installation_id`! I forgot, in this case, `app_id` is just referring to our internal DB id `github_apps.id`, not the GitHub App's external id `github_apps.app_id`. So as long as if they were from different instances, even if they share an external app id, they'd still have distinct internal DB ids.

**Comment by courier-new** [comment]
> I am noticing with multiple installs, the 2nd nodes "View in GitHub" link is broken. Do you see this on your end?

Oh hmm, I don't know how thoroughly I looked at that. I love your test org by the way. 😂 I'll look into that. Is the account that you're signed into on GitHub when you opened those links an owner of that org? It might 404 if your account isn't allowed to administer that installation.

**Comment by st0nebraker** [comment]
> Oh hmm, I don't know how thoroughly I looked at that. I love your test org by the way. 😂 I'll look into that. Is the account that you're signed into on GitHub when you opened those links an owner of that org? It might 404 if your account isn't allowed to administer that installation.

Haha thanks ♀️

Oh I'm not the owner of that org, I believe it allowed me to sign a commit through that org though with this setup. Let me test again, I have to step away but will test through this again when I'm back. Even so, we should maybe add more messaging around this edge case, right?

**Comment by courier-new** [comment]
> Oh I'm not the owner of that org, I believe it allowed me to sign a commit through that org though with this setup.

Yeah that would be expected to me. A normal user should still be able authenticate with the installation, they might just not be able to administer it (and hence might not be able to see the install page for it on GitHub)

> Even so, we should maybe add more messaging around this edge case, right?

Probably! But we won't be able to know if the current user is an org owner or not. We have this note on the "Creation" page about installing apps on orgs: "Only [organization owners](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization#organization-owners) can register GitHub Apps."

Maybe I could add a note like that at the top of the "Installations" list? Like "Only [organization owners](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization#organization-owners) can manage installations for organizations."?

**Comment by st0nebraker** [comment]
> Maybe I could add a note like that at the top of the "Installations" list? Like "Only [organization owners](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization#organization-owners) can manage installations for organizations."?

That sounds like a great solution to me! I just tested again (video below) with 3 orgs, one I'm just a member of, not an owner. The org settings aren't available, but everything else from SG batches commit signing works well. I think this disclaimer would be 👌🏻 


https://github.com/sourcegraph/sourcegraph/assets/59381432/d3b41123-b4de-4719-b992-b8f8a902192a

**Comment by courier-new** [comment]
Great, thank you for testing and confirming that. 😄 I'll follow up on that.

File: migrations/frontend/squashed.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (pjlast) [reply migrations/frontend/squashed.sql:5574]: Ah okay, I should probably have checked what `app_id` was referring to here, just automatically thought it was referring to the GitHub App ID on GitHub itself. Okay cool! All good then!
